timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -403,28 +361,60 @@ void __global__ k_set_sampled_weight(
     __syncthreads();
     block_energy_reduce<THREADS_PER_BLOCK>(accumulators, threadIdx.x);
     if (threadIdx.x == 0) {
-        log_weights[sample_idx] =
-            fixed_point_overflow(accumulators[0]) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(accumulators[0]);
+        intermediate_accum[blockIdx.x] = accumulators[0];
     }
 }
 
-template void __global__ k_set_sampled_weight<float, 512>(
+template void __global__ k_set_sampled_weight_block<float, 512>(
     const int N,
     const int mol_size,
-    const int *__restrict__ samples, // [1]
     const int *__restrict__ target_atoms,
-    const int *__restrict__ mol_offsets,
     const float *__restrict__ per_atom_energies,
     const float inv_kT,
-    float *__restrict__ log_weights);
-template void __global__ k_set_sampled_weight<double, 512>(
+    __int128 *__restrict__ intermediate_accum);
+template void __global__ k_set_sampled_weight_block<double, 512>(
     const int N,
     const int mol_size,
-    const int *__restrict__ samples, // [1]
     const int *__restrict__ target_atoms,
-    const int *__restrict__ mol_offsets,
     const double *__restrict__ per_atom_energies,
     const double inv_kT,
+    __int128 *__restrict__ intermediate_accum);
+
+template <typename RealType, int THREADS_PER_BLOCK>
+void __global__ k_set_sampled_weight_reduce(
+    const int intermediates,
+    const int *__restrict__ samples, // [1]
+    const __int128 *__restrict__ intermediate_accum,
+    RealType *__restrict__ log_weights) {
+    __shared__ __int128 accumulators[THREADS_PER_BLOCK];
+
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    // Zero all of the accumulators
+    __int128 accumulator = 0;
+    while (idx < intermediates) {
+        accumulator += intermediate_accum[idx];
+        idx += gridDim.x * blockDim.x;
+    }
+    accumulators[threadIdx.x] = accumulator;
+    __syncthreads();
+    block_energy_reduce<THREADS_PER_BLOCK>(accumulators, threadIdx.x);
+    if (threadIdx.x == 0) {
+        int sample_idx = samples[0];
+        log_weights[sample_idx] =
+            fixed_point_overflow(accumulators[0]) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(accumulators[0]);
+    }
+}
+
+template void __global__ k_set_sampled_weight_reduce<float, 512>(
+    const int intermediates,
```

Comment:
rename to `num_intermediates` and document the size?

Reply:
Done in https://github.com/proteneer/timemachine/pull/1225/commits/8caa3f77dfc2dcc87b2411cc4dde881b4111d464